src/Arcus.Messaging.Core/MessageHandling/MessageHandler.cs

```diff
@@ -237,15 +237,8 @@ private static ProcessMessageAsync DetermineMessageImplementation<TMessage, TMes
         {
             return async (rawMessage, generalMessageContext, correlationInfo, cancellationToken) =>
             {
-                if (rawMessage is not TMessage message)
-                {
-                    return MessageProcessingResult.Failure(generalMessageContext.MessageId, MessageProcessingError.MatchedHandlerFailed, $"requires message type {typeof(TMessage).Name}");
-                }
-
-                if (generalMessageContext is not TMessageContext messageContext)
-                {
-                    return MessageProcessingResult.Failure(generalMessageContext.MessageId, MessageProcessingError.MatchedHandlerFailed, $"requires message context type {typeof(TMessageContext).Name}");
-                }
+                var message = (TMessage) rawMessage;
```

Comment:
Isn't it possible that this results in a cast exception ?
(Previous code checked if the received message was of type `TMessage` ?

Reply:
It was moved to another place. The `MessageRouter` always first checks if it matches before going through this one, if it would be added here as well it would be duplication, but I agree, it is not the best design-wise.
Let's see if we can make this more rational..